### PR TITLE
Make concatenation of SubString{AnnotatedString} preserve annotations

### PR DESCRIPTION
**Pull request by tecosaur**
`SubStrings` have been overlooked, and thanks to a few compiler quirks (relating to inlining and effect analysis), adding support for them is unfortunately a little more complicated than adding a `|| s isa SubString{<:AnnotatedString}` clause thanks to the new generated runtime-checks.

To maintain the zero-overhead non-annotated code path, we need to implement a separate function `_isannotated`, which we also make use of to simplify the current join method.

### CODE REVIEW

**Comment by tecosaur** [comment]
I must admit, I can't see what's going on with CI here.

**Comment by tecosaur** [comment]
Just pushed the use of `_isannotated` to simplify `lpad` and `rpad`.

**Comment by tecosaur** [comment]
The CI failure seems unrelated:

```
make[1]: *** [stdlib/libLLVM_jll.release.image] Terminated: 15
Terminated: 15
🚨 Error: The command exited with status 143
```

**Comment by tecosaur** [comment]
Now with tests :)

**Comment by tecosaur** [comment]
Ok, what's going on with CI? It seems consistently on the fritz again.

**Comment by tecosaur** [comment]
I'd think this to be a pretty straightforward small improvement that could be quickly merged, is there anything holding this back?

**Comment by tecosaur** [comment]
It would be great if somebody with merge rights could spare the time to have a look at this PR.

**Review by LilithHafner — APPROVED** [review]
This construction
```julia
if annotated
    annotatedstring
else
    string
end
```
doesn't feel natural in the generic-function/multiple-dispatch paradigm, which makes me averse to touching this. However, that aversion does not block this PR as this PR simply improves the situation by shifting some existing hardcoded || statements into generic functions (in order fix some substring functionality)

**Comment by tecosaur** [comment]
I get that aversion. However, the reason why "if annotated" now pops up in a few places is because it's extended a status quo of "hardcoded `String`". I'm not sure what a more idiomatic-looking (and non-breaking) approach would look like, and nobody really suggested one during the original review of this feature.

I do recall Valentin (Churavy) suggesting that `annotatedstring` should be merged with `string` (which I'd personally be quite happy with), but IIRC one or two other people in triage veto'd that.

**Pull request merged**
